services/intersection-api/api/src/main/java/us/dot/its/jpo/ode/api/controllers/ReportController.java

```diff
@@ -0,0 +1,89 @@
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.data.mongodb.core.query.Query;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.security.access.prepost.PreAuthorize;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import us.dot.its.jpo.ode.api.accessors.reports.ReportRepository;
+import us.dot.its.jpo.ode.api.models.ReportDocument;
+import us.dot.its.jpo.ode.api.services.ReportService;
+
+@Slf4j
+@RestController
+@ConditionalOnProperty(name = "enable.api", havingValue = "true", matchIfMissing = false)
+public class ReportController {
+
+    @Autowired
+    ReportService reportService;
+
+    @Autowired
+    ReportRepository reportRepo;
+
+    @RequestMapping(value = "/reports/generate", method = RequestMethod.GET, produces = "application/octet-stream")
+    @PreAuthorize("@PermissionService.isSuperUser() || (@PermissionService.hasIntersection(#intersectionID) and @PermissionService.hasRole('USER')) ")
+    public byte[] generateReport(
+            @RequestParam(name = "intersection_id") int intersectionID,
+            @RequestParam(name = "road_regulator_id", required = false) Integer roadRegulatorID,
+            @RequestParam(name = "start_time_utc_millis") long startTime,
+            @RequestParam(name = "end_time_utc_millis") long endTime) {
+        log.debug("Generating Report");
+
+        if (roadRegulatorID == null) {
+            roadRegulatorID = -1;
+        }
+
+        ReportDocument document = reportService.buildReport(intersectionID, roadRegulatorID.toString(), startTime,
+                endTime);
+
+        return document.getReportContents();
+    }
+
+    @RequestMapping(value = "/reports/list", method = RequestMethod.GET, produces = "application/json")
+    @PreAuthorize("@PermissionService.isSuperUser() || @PermissionService.hasRole('USER')")
+    public ResponseEntity<List<ReportDocument>> listReports(
+            @RequestParam(name = "report_name", required = false) String reportName,
+            @RequestParam(name = "intersection_id", required = false) int intersectionID,
+            @RequestParam(name = "road_regulator_id", required = false) Integer roadRegulatorID,
+            @RequestParam(name = "start_time_utc_millis") long startTime,
+            @RequestParam(name = "end_time_utc_millis") long endTime,
+            @RequestParam(name = "latest") boolean latest) {
+
+        if (roadRegulatorID == null) {
+            roadRegulatorID = -1;
+        }
+
+        Query query = reportRepo.getQuery(reportName, intersectionID, roadRegulatorID, startTime, endTime,
+                false,
+                latest);
+        return ResponseEntity.ok(reportRepo.find(query));
```

Comment:
issue(blocking): executing query second time for logging purposes

Reply:
Extra query has been removed